### PR TITLE
fix(stm32h7): timer version assignements

### PR DESCRIPTION
**Pull request by elagil**
Closes https://github.com/embassy-rs/stm32-data/issues/509

### CODE REVIEW

**Comment by elagil** [comment]
Apparently, multiple timer versions are not allowed at once. I am quite sure that the H7 parts do actually use timers of different versions.

For example, I checked the `TISEL` `byte_offset` field, which for the basic timers on the H7 indicates `timer_v1` (the offset is indeed 104, instead of 92 for `timer_v2`).

**Comment by elagil** [comment]
I made a "fix" to allow mismatching timer peripheral versions in the chip generator.

File: stm32-metapac-gen/src/lib.rs [hunk 1]
```diff
@@ -72,7 +72,7 @@ impl Gen {
 
             if let Some(bi) = &p.registers {
                 if let Some(old_version) = peripheral_versions.insert(bi.kind.clone(), bi.version.clone()) {
-                    if old_version != bi.version {
+                    if (old_version != bi.version) && (bi.kind != "timer") {
```
Comment (Dirbaio) [inline stm32-metapac-gen/src/lib.rs:75]: this won't work when you try to actually use this with embassy-stm32.

If H7 neither matches v1 nor v2, the fix is to make a new version and make all h7 timers use that.

Comment (elagil) [reply stm32-metapac-gen/src/lib.rs:75]: Ok, I understand. I will do that then.

**Comment by elagil** [comment]
@Dirbaio I created a single file in the following way:
- Copy `timer_v2.yaml` to `timer_h7.yaml`.
- Remove all definitions from `timer_h7.yaml` that are the exactly same in `timer_v1.yaml`.
- Rename any remaining definitions (`enums`, `fieldsets`, `blocks`) to end in `_V2`, as well as their references.
- Copy all definitions from `timer_v1.yaml` into `timer_h7.yaml` (to sensible locations)

Is there a good way to test this?

**Comment by elagil** [comment]
`embassy-stm32::timer` does not yet support multiple driver versions at once. For example, `set_trigger_source` only accepts an input of type `TriggerSource`, but there now has to be an additional `TriggerSourceV2` type for the V2 timers of the chip.

@Dirbaio Do you have a suggestion for avoiding a lot of code duplication? Of course, we can just compile for either the V1 or the V2 timers. The others would not be usable in that case (not great).

**Comment by elagil** [comment]
Is this a good approach, or should we change it?

**Comment by Dirbaio** [comment]
@elagil if you look through the diff the CI bot posted, there's many timers where the `registers` field is gone. This will make those be unusable with the HAL. The reason for it is no longer matching any rule in `perimap`. Can you take a look?

**Comment by elagil** [comment]
@Dirbaio I know, the disabled timers are the v1 timers. This PR deliberately makes it so that only v2 timers are active.

Up to now, all timers (even v2) were used with the v1 register set. I chose to enable only v2 timers with this PR because there are more v2 than v1 timers, and they have more features. These now use the v2 register definitions, as they should.

Do you have an idea for solving this problem? v1 and v2 do not have compatible registers and cannot be merged into a single definition file. As you said, in embassy, we cannot use multiple driver versions at once.

**Comment by Dirbaio** [comment]
ah sorry I should've read the thread instead of assuming my memory is in working order :sweat_smile: 

There's people out there using a lot of timers, sometimes someone asks in Matrix how to free the timer used by the time driver because they're running out. I can imagine making a bunch of timers not have registers will be unpopular. I'd say allowing all timers to be usable is better than having 100% accuracy but only allowing using half the timers.

so yes, if some timers have some features and the others don't then we'll need two copies of fieldsets/enums and make the right timer block use the right ones...

The HAL will have to deal with this ... somehow. Perhaps it's okay for now to be slightly inaccurate, for example using the "more advanced" enums/fieldsets for all timers and hope users read the docs of which features are available on which timers.

**Comment by elagil** [comment]
> The HAL will have to deal with this ... somehow.

Ok, I will think of something for handling two versions at once without excessive duplication. I need some of the advanced features of the v2 timers (e.g. USB SOF signal as trigger), so I can't just use v1 definitions.

**Comment by kalkyl** [comment]
Cool, just curious :)

**Comment by elagil** [comment]
STM32H7 does not use v2 timers, as previously assumed.
It uses TS enum entries that were not included in the v1 definition yet.

**Pull request merged**